filebeat/crawler/prospector.go

```diff
@@ -28,14 +27,13 @@ type Prospectorer interface {
 	Run()
 }
 
-func NewProspector(cfg *common.Config, registrar *Registrar, spoolerChan chan *input.FileEvent) (*Prospector, error) {
+func NewProspector(cfg *common.Config, states input.States, spoolerChan chan *input.FileEvent) (*Prospector, error) {
 	prospector := &Prospector{
 		config:        defaultConfig,
-		registrar:     registrar,
 		spoolerChan:   spoolerChan,
 		harvesterChan: make(chan *input.FileEvent),
 		done:          make(chan struct{}),
-		states:        input.NewStates(),
+		states:        &states,
```

Comment:
states is defined like this:

```
 // States handles list of FileState
 type States struct {
 	states []FileState
 	mutex  sync.Mutex
 }
```

The copy of 'states' is mostly shallow. Some code checkers (newer `go vet` should) will complain about copy on mutex. The underlying byte-buffer is shared when slice is copied. The `(*States).Update` method can override a file its state, if file is member of slice. Due to mutex being copied, any 2 go-routines attempting to updates a file state  for the same file might actually race.